projectq/ops/_qubit_operator.py

```diff
@@ -226,6 +251,141 @@ def isclose(self, other, rel_tol=1e-12, abs_tol=1e-12):
+            qubits: one Qubit object, one list of Qubit objects, one Qureg
+                    object, or a tuple of the former three cases.
+
+        Raises:
+            TypeError: If QubitOperator is not unitary or applied to more than
+                       one quantum register.
+            ValueError: If quantum register does not have enough qubits
+        """
+        # Check that input is only one qureg or one qubit
+        qubits = self.make_tuple_of_qureg(qubits)
+        if len(qubits) != 1:
+            raise TypeError("Only one qubit or qureg allowed.")
+        # Check that operator is unitary
+        if not len(self.terms) == 1:
+            raise TypeError("Only unitary QubitOperators can be applied to "
+                            "qubits.")
+        (term, coefficient), = self.terms.items()
+        phase = cmath.phase(coefficient)
+        if (abs(coefficient) < 1 - EQ_TOLERANCE or
+                abs(coefficient) > 1 + EQ_TOLERANCE):
+            raise TypeError("Only unitary QubitOperators can be applied to "
+                            "qubits.")
+        # Test if we need to apply only Ph
+        if term == ():
+            Ph(phase) | qubits[0][0]
+            return
+        # Check that Qureg has enough qubits:
+        num_qubits = len(qubits[0])
+        non_trivial_qubits = set()
+        for index, action in term:
+            non_trivial_qubits.add(index)
+        if max(non_trivial_qubits) >= num_qubits:
+            raise ValueError("QubitOperator acts on more qubits than the gate "
+                             "is applied to.")
+        # Apply X, Y, Z, if QubitOperator acts only on one qubit
+        if len(term) == 1:
+            if term[0][1] == "X":
+                X | qubits[0][term[0][0]]
+            elif term[0][1] == "Y":
+                Y | qubits[0][term[0][0]]
+            elif term[0][1] == "Z":
+                Z | qubits[0][term[0][0]]
+            Ph(phase) | qubits[0][term[0][0]]
+            return
+        # Create new QubitOperator gate with rescaled qubit indices in
+        # 0,..., len(non_trivial_qubits) - 1
+        new_index = dict()
+        non_trivial_qubits = sorted(list(non_trivial_qubits))
+        for i in range(len(non_trivial_qubits)):
+            new_index[non_trivial_qubits[i]] = i
+        new_qubitoperator = QubitOperator()
+        assert len(new_qubitoperator.terms) == 0
+        new_term = tuple([(new_index[index], action)
+                          for index, action in term])
+        new_qubitoperator.terms[new_term] = coefficient
+        new_qubits = [qubits[0][i] for i in non_trivial_qubits]
+        # Apply new gate
+        cmd = new_qubitoperator.generate_command(new_qubits)
+        apply_command(cmd)
+
+    def get_inverse(self):
+        """
+        Return the inverse gate if QubitOperator is unitary.
+
+        Raises:
+            NotInvertible: inverse is not implemented if not unitary
+        """
+        if len(self.terms) == 1:
```

Comment:
and abs(coefficient)...

Reply:
done

Reply:
It won't raise an exception now, though...